### PR TITLE
Optimize bit reversal

### LINKED ISSUE #444: Use Plonky2's bit reverse implementation
Jacqui's bit reverse implementation in Plonky2 is faster than the naive implementation we currently have; we should port it across and use it.

### PR DESCRIPTION
**Pull request by huitseeker**
This imports the bit reversal implementation from Plonky2, focusing on `reverse_slice_index_bits`. FIxes #444.
This also imports some unit tests on the edge cases of `log2_strict_usize`, which were in the same file.

I have not tuned `BIG_T_SIZE` and `SMALL_ARR_SIZE` currently set for a rather conservative 64KB L1.
Values for bench machines are for M3 Max:
```
huitseeker@binky➜tmp/plonky3/util(add_tests)» sysctl -a | grep -Ee 'hw\.l[12].*?cachesize'                                                                                                                                          
hw.l1icachesize: 131072
hw.l1dcachesize: 65536
hw.l2cachesize: 4194304
```
For Ryzen 9950X:
```
huitseeker@sleipnir➜huitseeker/tmp/plonky3(add_tests)» lscpu | grep "cache"                                                                                                                                                        
L1d cache:                            768 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             16 MiB (16 instances)
L3 cache:                             64 MiB (2 instances)
```
(the later compiled with `target-cpu=native`)

~~As the benches are mitigated on performance (considering the priority of x86_64), I'm happy to revert to the more Pareto-efficient 1419ae529dc77decad50cfa70e60afbbf72af9f4.~~ (see [this comment](https://github.com/Plonky3/Plonky3/pull/596#issuecomment-2598437978))

<details><summary>Benches on 1419ae529dc77decad50cfa70e60afbbf72af9f4 (toggle me!) </summary>

**x86_64 benches (Ryzen 9 9950X)**:

```
reverse_bits/2          time:   [554.71 ps 554.73 ps 554.76 ps]
                        change: [+0.0497% +0.0697% +0.0897%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/8          time:   [555.11 ps 555.22 ps 555.37 ps]
                        change: [+5.0627% +5.1082% +5.1658%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/32         time:   [530.03 ps 530.37 ps 530.77 ps]
                        change: [+0.1313% +0.1753% +0.2252%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/256        time:   [554.59 ps 554.62 ps 554.65 ps]
                        change: [+4.9471% +4.9616% +4.9781%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/65536      time:   [554.67 ps 554.70 ps 554.73 ps]
                        change: [+4.5514% +4.5659% +4.5803%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/16777216   time:   [529.43 ps 530.02 ps 530.60 ps]
                        change: [-4.9272% -4.8555% -4.7821%] (p = 0.00 < 0.05)
                        Performance has improved.

reverse_slice_index_bits/2
                        time:   [7.7681 ns 7.7864 ns 7.8056 ns]
                        change: [-0.0547% +0.2431% +0.5594%] (p = 0.14 > 0.05)
                        No change in performance detected.
reverse_slice_index_bits/8
                        time:   [9.3444 ns 9.3565 ns 9.3694 ns]
                        change: [-4.7702% -4.5714% -4.3713%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/32
                        time:   [16.041 ns 16.047 ns 16.053 ns]
                        change: [-15.261% -15.012% -14.785%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/256
                        time:   [98.456 ns 98.489 ns 98.524 ns]
                        change: [-47.381% -47.127% -46.867%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/65536
                        time:   [205.94 µs 205.97 µs 205.99 µs]
                        change: [-0.3076% -0.2808% -0.2556%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Benchmarking reverse_slice_index_bits/16777216: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 16.4s, or reduce sample count to 30.
reverse_slice_index_bits/16777216
                        time:   [162.85 ms 163.16 ms 163.53 ms]
                        change: [-3.3245% -3.1013% -2.8419%] (p = 0.00 < 0.05)
                        Performance has improved.
```

**aarch64 benches (M3 Max)**:

```
reverse_bits/2          time:   [649.86 ps 654.88 ps 658.72 ps]
                        change: [-1.6618% -0.9119% -0.2410%] (p = 0.01 < 0.05)
                        Change within noise threshold.
reverse_bits/8          time:   [657.96 ps 660.17 ps 662.11 ps]
                        change: [+0.0576% +0.3996% +0.7267%] (p = 0.02 < 0.05)
                        Change within noise threshold.
reverse_bits/32         time:   [658.14 ps 660.56 ps 662.78 ps]
                        change: [-0.2617% +0.1755% +0.6037%] (p = 0.43 > 0.05)
                        No change in performance detected.
reverse_bits/256        time:   [606.35 ps 619.43 ps 632.47 ps]
                        change: [+13.039% +14.560% +16.052%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/65536      time:   [578.02 ps 589.20 ps 601.88 ps]
                        change: [+8.9996% +10.968% +12.852%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/16777216   time:   [555.35 ps 556.69 ps 558.12 ps]
                        change: [-4.8527% -3.3453% -1.9049%] (p = 0.00 < 0.05)
                        Performance has improved.

reverse_slice_index_bits/2
                        time:   [15.511 ns 15.543 ns 15.576 ns]
                        change: [+1.7267% +1.9085% +2.0940%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/8
                        time:   [21.351 ns 21.396 ns 21.441 ns]
                        change: [+0.1883% +0.5824% +0.9700%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_slice_index_bits/32
                        time:   [35.606 ns 35.680 ns 35.759 ns]
                        change: [+10.806% +11.128% +11.422%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/256
                        time:   [200.90 ns 201.90 ns 202.83 ns]
                        change: [+40.215% +40.784% +41.378%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/65536
                        time:   [144.72 µs 144.96 µs 145.18 µs]
                        change: [+2.8216% +3.3257% +3.8254%] (p = 0.00 < 0.05)
                        Performance has regressed.
Benchmarking reverse_slice_index_bits/16777216: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.8s, or reduce sample count to 40.
reverse_slice_index_bits/16777216
                        time:   [107.27 ms 107.65 ms 108.07 ms]
                        change: [+0.6871% +1.4264% +2.1466%] (p = 0.00 < 0.05)
                        Change within noise threshold.
```
</details>

<details><summary>Benches on bc711ef22cd342dcf007ccb7031d201789a3ee1d (toggle me!)</summary>

**x86 benches (Ryzen 9 9950X):**

```
reverse_bits/2          time:   [554.25 ps 554.27 ps 554.30 ps]
                        change: [+0.2328% +0.2456% +0.2576%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/8          time:   [554.39 ps 554.41 ps 554.44 ps]
                        change: [+4.7186% +4.7534% +4.7817%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/32         time:   [554.67 ps 554.69 ps 554.72 ps]
                        change: [+4.7319% +4.7471% +4.7608%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/256        time:   [529.74 ps 529.76 ps 529.78 ps]
                        change: [+0.1329% +0.1951% +0.2496%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/65536      time:   [552.72 ps 553.38 ps 553.94 ps]
                        change: [-1.7682% -1.4123% -0.9860%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/16777216   time:   [554.71 ps 554.74 ps 554.78 ps]
                        change: [+0.1098% +0.1230% +0.1354%] (p = 0.00 < 0.05)
                        Change within noise threshold.

reverse_slice_index_bits/2
                        time:   [7.7281 ns 7.7405 ns 7.7534 ns]
                        change: [+1.2297% +1.7211% +2.2094%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/8
                        time:   [9.1964 ns 9.2168 ns 9.2380 ns]
                        change: [-6.5340% -6.3151% -6.0937%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/32
                        time:   [16.022 ns 16.030 ns 16.040 ns]
                        change: [-17.666% -17.372% -17.082%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/256
                        time:   [97.458 ns 97.495 ns 97.536 ns]
                        change: [-45.120% -44.876% -44.633%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/65536
                        time:   [200.46 µs 200.56 µs 200.63 µs]
                        change: [+0.9489% +0.9977% +1.0456%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Benchmarking reverse_slice_index_bits/16777216: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 29.1s, or reduce sample count to 10.
reverse_slice_index_bits/16777216
                        time:   [268.27 ms 269.06 ms 269.92 ms]
                        change: [+65.295% +65.787% +66.340%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

**aarch64 benches (M3 Max)**:

```
reverse_bits/2          time:   [658.53 ps 659.63 ps 660.85 ps]
                        change: [-0.3605% +0.2031% +0.8141%] (p = 0.52 > 0.05)
                        No change in performance detected.
reverse_bits/8          time:   [657.96 ps 658.89 ps 659.89 ps]
                        change: [-0.4917% -0.1500% +0.2002%] (p = 0.40 > 0.05)
                        No change in performance detected.
reverse_bits/32         time:   [651.35 ps 656.26 ps 659.17 ps]
                        change: [-0.9491% -0.5186% -0.0554%] (p = 0.02 < 0.05)
                        Change within noise threshold.
reverse_bits/256        time:   [639.64 ps 648.70 ps 655.37 ps]
                        change: [-3.6259% -1.7284% +0.2640%] (p = 0.09 > 0.05)
                        No change in performance detected.
reverse_bits/65536      time:   [621.40 ps 631.53 ps 639.65 ps]
                        change: [+0.1756% +2.4584% +4.7819%] (p = 0.04 < 0.05)
                        Change within noise threshold.
reverse_bits/16777216   time:   [641.48 ps 647.11 ps 651.37 ps]
                        change: [-2.1677% +0.1160% +2.3395%] (p = 0.93 > 0.05)
                        No change in performance detected.

reverse_slice_index_bits/2
                        time:   [15.593 ns 15.644 ns 15.706 ns]
                        change: [+1.1210% +1.6573% +2.1595%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/8
                        time:   [21.113 ns 21.153 ns 21.203 ns]
                        change: [+0.2235% +0.7705% +1.2916%] (p = 0.01 < 0.05)
                        Change within noise threshold.
reverse_slice_index_bits/32
                        time:   [33.069 ns 33.174 ns 33.311 ns]
                        change: [+2.0886% +2.4791% +2.8463%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/256
                        time:   [150.98 ns 152.05 ns 153.30 ns]
                        change: [+4.3236% +4.9794% +5.6499%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/65536
                        time:   [120.80 µs 121.13 µs 121.46 µs]
                        change: [-15.077% -14.663% -14.249%] (p = 0.00 < 0.05)
                        Performance has improved.
Benchmarking reverse_slice_index_bits/16777216: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.3s, or reduce sample count to 70.
reverse_slice_index_bits/16777216
                        time:   [62.464 ms 62.634 ms 62.829 ms]
                        change: [-41.714% -41.338% -40.973%] (p = 0.00 < 0.05)
                        Performance has improved.
```

</details>

### CODE REVIEW

File: util/Cargo.toml [hunk 1]
```diff
@@ -6,3 +6,11 @@ license = "MIT OR Apache-2.0"
...
 [dependencies]
 serde.workspace = true
+
+[dev-dependencies]
+rand.workspace = true
+criterion.workspace = true
+
+[[bench]]
+name = "bit_reverse"
+harness = false
```
Comment (unzvfu) [inline util/Cargo.toml:16]: Nit: Missing newline at end of file.

**Review by unzvfu — CHANGES_REQUESTED** [review]
Thanks for this contribution. I would strongly prefer to retain the extensive inline comments and function-documentation comments from the original, without which this code becomes much harder to understand.

I also note that you've changed all the unsafe calls
```
swap(arr.get_unchecked_mut(src), arr.get_unchecked_mut(dst))
```
with the equivalent safe calls
```
arr.swap(src, dst)
```
Did you benchmark the difference? I haven't checked this code specifically, but the compiler can't always elide the check in the safe indexing calls which can be really expensive.

**Comment by unzvfu** [comment]
> As the benches are mitigated on performance (considering the priority of x86_64), I'm happy to revert to the more Pareto-efficient https://github.com/Plonky3/Plonky3/commit/1419ae529dc77decad50cfa70e60afbbf72af9f4.

Regarding this, if we must choose, I think we should err towards whatever is fastest on `x86_64` for now. It would be nice to have the fastest times on both architectures, but that would probably complicate the implementation more than is really worthwhile, since none of these functions are bottlenecks iirc.

**Comment by huitseeker** [comment]
Rebased, re-added comments, and re-added calls to unchecked. The performance picture is quite similar.

<details><Summary> Benchmarks on 373bbd3 (toggle me!)</summary>

**on x86_64 (Ryzen 9 9950X)**:
```
reverse_bits/2          time:   [556.12 ps 556.28 ps 556.53 ps]
                        change: [+0.3869% +0.4483% +0.5264%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/8          time:   [556.39 ps 556.42 ps 556.45 ps]
                        change: [+0.3291% +0.3448% +0.3573%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/32         time:   [556.47 ps 556.52 ps 556.58 ps]
                        change: [+4.8394% +4.8591% +4.8778%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_bits/256        time:   [557.27 ps 557.61 ps 558.03 ps]
                        change: [+0.4748% +0.5277% +0.5756%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/65536      time:   [556.22 ps 556.26 ps 556.30 ps]
                        change: [+0.2791% +0.3012% +0.3195%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/16777216   time:   [556.32 ps 556.36 ps 556.41 ps]
                        change: [+5.2723% +5.3090% +5.3383%] (p = 0.00 < 0.05)
                        Performance has regressed.

reverse_slice_index_bits/2
                        time:   [7.7353 ns 7.7552 ns 7.7753 ns]
                        change: [-1.3058% -0.9487% -0.6066%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_slice_index_bits/8
                        time:   [9.0896 ns 9.1187 ns 9.1483 ns]
                        change: [-7.5568% -7.2836% -7.0244%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/32
                        time:   [14.322 ns 14.335 ns 14.348 ns]
                        change: [-27.474% -27.316% -27.173%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/256
                        time:   [99.739 ns 99.780 ns 99.832 ns]
                        change: [-42.822% -42.588% -42.352%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/65536
                        time:   [191.90 µs 191.97 µs 192.06 µs]
                        change: [-6.4595% -6.3642% -6.2548%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/16777216
                        time:   [271.77 ms 272.23 ms 272.69 ms]
                        change: [+65.384% +65.769% +66.136%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

**on aarch64 (M3 Max)**:
```
reverse_bits/2          time:   [635.06 ps 642.43 ps 649.39 ps]
                        change: [-5.3831% -4.3877% -3.2902%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_bits/8          time:   [643.66 ps 651.13 ps 657.24 ps]
                        change: [-2.2598% -1.5755% -0.8841%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/32         time:   [644.55 ps 652.55 ps 660.07 ps]
                        change: [-2.4090% -1.5073% -0.5039%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/256        time:   [610.10 ps 622.17 ps 633.72 ps]
                        change: [-0.5575% +1.8662% +4.0999%] (p = 0.12 > 0.05)
                        No change in performance detected.
reverse_bits/65536      time:   [610.38 ps 622.11 ps 633.73 ps]
                        change: [+0.0314% +2.2721% +4.6853%] (p = 0.06 > 0.05)
                        No change in performance detected.
reverse_bits/16777216   time:   [602.76 ps 616.13 ps 627.97 ps]
                        change: [-4.6680% -2.5346% -0.3219%] (p = 0.03 < 0.05)
                        Change within noise threshold.

reverse_slice_index_bits/2
                        time:   [15.520 ns 15.565 ns 15.614 ns]
                        change: [+0.7651% +1.2620% +1.7197%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_slice_index_bits/8
                        time:   [21.077 ns 21.135 ns 21.196 ns]
                        change: [-1.1586% -0.6395% -0.1422%] (p = 0.02 < 0.05)
                        Change within noise threshold.
reverse_slice_index_bits/32
                        time:   [33.484 ns 33.694 ns 33.926 ns]
                        change: [+2.1416% +2.9553% +3.7610%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/256
                        time:   [149.33 ns 150.43 ns 151.56 ns]
                        change: [-0.7738% -0.1035% +0.5201%] (p = 0.76 > 0.05)
                        No change in performance detected.
reverse_slice_index_bits/65536
                        time:   [122.04 µs 122.40 µs 122.81 µs]
                        change: [-17.413% -17.029% -16.649%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/16777216
                        time:   [62.282 ms 62.579 ms 62.891 ms]
                        change: [-46.730% -46.177% -45.613%] (p = 0.00 < 0.05)
                        Performance has improved.
```

</details>

**Comment by unzvfu** [comment]
> Rebased, re-added comments, and re-added calls to unchecked. The performance picture is quite similar.
> Benchmarks on [373bbd3](https://github.com/Plonky3/Plonky3/commit/373bbd33c7a8c75467be0a51b06b0511b4313d6a) (toggle me!)

Okay, thanks for that. Since using unchecked doesn't affect performance I agree that your use of `swap` is preferable. I think this will be ready to merge once you revert back to using `swap`, and (as mentioned previously) revert to the version that is fastest on x86.

**Comment by huitseeker** [comment]
@unzvfu I'm a dunce: in my original implementation of the full reversal (see commit "refactor: implement the full bit reversal"), I benched on x86 that the performance was downgraded on large inputs, but failed to notice that in trying to slice and dice the algorithmic improvements in separate layers, I had left one item on my TODO-list: the optimized in-place transposition. That commit message was therefore a lie, it was the "not-quite-full" bit reversal.

When I do add the transposition (as I did as part of my last commit), the x86 benches look a lot better.

<details><summary> Benchmarks on 51b542 </summary>

**x86_64 benches (Ryzen 9 9950X):**
```
reverse_bits/2          time:   [530.25 ps 530.28 ps 530.31 ps]
                        change: [-2.5636% -2.1497% -1.7399%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_bits/8          time:   [528.44 ps 528.51 ps 528.56 ps]
                        change: [-5.1042% -5.0565% -5.0107%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_bits/32         time:   [531.43 ps 531.72 ps 532.09 ps]
                        change: [+0.2244% +0.2929% +0.3613%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/256        time:   [554.87 ps 554.90 ps 554.92 ps]
                        change: [-0.1797% -0.1666% -0.1535%] (p = 0.00 < 0.05)
                        Change within noise threshold.
reverse_bits/65536      time:   [549.62 ps 551.28 ps 552.55 ps]
                        change: [-2.9009% -2.4647% -1.9774%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_bits/16777216   time:   [554.43 ps 554.46 ps 554.49 ps]
                        change: [-0.2422% -0.2273% -0.2118%] (p = 0.00 < 0.05)
                        Change within noise threshold.

reverse_slice_index_bits/2
                        time:   [8.1166 ns 8.1303 ns 8.1456 ns]
                        change: [+8.2811% +8.5809% +8.8844%] (p = 0.00 < 0.05)
                        Performance has regressed.
reverse_slice_index_bits/8
                        time:   [8.8556 ns 8.8719 ns 8.8891 ns]
                        change: [-10.115% -9.9588% -9.8056%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/32
                        time:   [15.128 ns 15.137 ns 15.147 ns]
                        change: [-20.142% -19.954% -19.771%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/256
                        time:   [97.962 ns 97.978 ns 97.996 ns]
                        change: [-46.546% -46.324% -46.110%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/65536
                        time:   [21.052 µs 21.054 µs 21.057 µs]
                        change: [-89.823% -89.819% -89.816%] (p = 0.00 < 0.05)
                        Performance has improved.
reverse_slice_index_bits/16777216
                        time:   [41.501 ms 41.547 ms 41.605 ms]
                        change: [-75.273% -75.244% -75.205%] (p = 0.00 < 0.05)
                        Performance has improved.
```
</details>

I think this now contains the full kit and kaboodle, and has much more positive results!

**Review by unzvfu — APPROVED** [review]
Great, this looks good to go. Thanks!
